overlord/snapstate/aliasesv2.go

```diff
@@ -104,11 +140,207 @@ func applyAliasesChange(st *state.State, snapName string, prevStatus AliasesStat
+	}
+	var firstErr error
+	touched = make(map[string][]string)
+	retired = make(map[string][]string)
+	for snapName, snapst := range snapStates {
+		aliases := snapst.Aliases
+		info, err := snapst.CurrentInfo()
+		if err != nil {
+			if firstErr == nil {
+				firstErr = err
+			}
+			continue
+		}
+		autoAliases, err := AutoAliases(st, info)
+		if err != nil {
+			if firstErr == nil {
+				firstErr = err
+			}
+			continue
+		}
+		for alias, target := range autoAliases {
+			curTarget := aliases[alias]
+			if curTarget == nil || curTarget.Auto != target {
+				touched[snapName] = append(touched[snapName], alias)
+			}
+		}
+		for alias, target := range aliases {
+			if target.Auto != "" && autoAliases[alias] == "" {
+				retired[snapName] = append(retired[snapName], alias)
+			}
+		}
+	}
+	return touched, retired, firstErr
+}
+
+// refreshAliases applies the current snap-declaration aliases
+// considering which applications exist in info and produces new aliases
+// for the snap.
+func refreshAliases(st *state.State, info *snap.Info, curAliases map[string]*AliasTarget) (newAliases map[string]*AliasTarget, err error) {
+	autoAliases, err := AutoAliases(st, info)
+	if err != nil {
+		return nil, err
+	}
+
+	newAliases = make(map[string]*AliasTarget, len(autoAliases))
+	// apply the current auto-aliases
+	for alias, target := range autoAliases {
+		if info.Apps[target] == nil {
+			// not an existing app
+			continue
+		}
+		newAliases[alias] = &AliasTarget{Auto: target}
+	}
+
+	// carry over the current manual ones
+	for alias, curTarget := range curAliases {
+		if curTarget.Manual == "" {
+			continue
+		}
+		if info.Apps[curTarget.Manual] == nil {
+			// not an existing app
+			continue
+		}
+		newTarget := newAliases[alias]
+		if newTarget == nil {
+			newAliases[alias] = &AliasTarget{Manual: curTarget.Manual}
+		} else {
+			// alias is both manually setup but has an underlying auto-alias
+			newAliases[alias] = &AliasTarget{Manual: curTarget.Manual, Auto: newTarget.Auto}
+		}
+	}
+	return newAliases, nil
+}
+
+type AliasConflictError struct {
+	Snap      string
+	Alias     string
+	Reason    string
+	Conflicts map[string][]string
+}
+
+func (e *AliasConflictError) Error() string {
+	if len(e.Conflicts) != 0 {
+		errParts := []string{"cannot enable"}
+		first := true
+		for snapName, aliases := range e.Conflicts {
+			if !first {
+				errParts = append(errParts, "nor")
+			}
+			if len(aliases) == 1 {
+				errParts = append(errParts, fmt.Sprintf("alias %q", aliases[0]))
+			} else {
+				errParts = append(errParts, fmt.Sprintf("aliases %s", strutil.Quoted(aliases)))
+			}
+			if first {
+				errParts = append(errParts, fmt.Sprintf("for %q,", e.Snap))
+				first = false
+			}
+			errParts = append(errParts, fmt.Sprintf("already enabled for %q", snapName))
+		}
+		// TODO: add recommendation about what to do next
+		return strings.Join(errParts, " ")
+	}
+	return fmt.Sprintf("cannot enable alias %q for %q, %s", e.Alias, e.Snap, e.Reason)
+}
+
```

Comment:
the old code used a similar function also to check the snapName vs all aliases case, I can probably just use different code for that case if we want, interestingly if we want to reduce nested iterations I think we could iterate case by case over the shortest of testAliases (aka enabled)  vs snapst.Aliases